eoscompanion/routes.py

```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
+        '''
+        app_infos = []
+
+        for result in results:
+            error, info = result
+
+            if error is not None:
+                logging.error('Encountered error in getting app info: %s', error)
+                continue
+
+            app_infos.append(info)
+
+        callback(None, app_infos)
+
+    def _load_application_info_thunk(desktop_id):
+        '''A thunk used here since we cannot use lambdas in for loops.'''
+        def _internal(callback):
+            def _on_got_application_info(_, result):
+                '''Marshal the error and result into a tuple.'''
+                try:
+                    info = application_listing_from_app_info(
+                        EosCompanionAppService.finish_load_application_info(result)
+                    )
+                except GLib.Error as error:
+                    callback(error, None)
+                    return
+
+                callback(None, info)
+
+            application_id = desktop_id_to_app_id(desktop_id)
+            EosCompanionAppService.load_application_info(application_id,
+                                                         None,
+                                                         _on_got_application_info)
+
+        return _internal
+
+    desktop_ids = list(set([
+        desktop_id for desktop_id in yield_desktop_ids_from_feed_models(models)
+    ]))
+    all_asynchronous_function_calls_closure([
+        _load_application_info_thunk(desktop_id) for desktop_id in desktop_ids
+    ], _on_received_all_app_info_results)
+
+
+def sources_from_content_feed_models(models, query, callback):
+    '''Generate the "sources" entry for the feed JSON response.
+
+    This response is a summary of all the sources that were queried in
+    the feed. A source might generate multiple items to put in the feed
+    so we can save some space in the payload by having a unique source
+    identifier and specifying its detail once.
+
+    Sources list generation happens asynchronously as it requires
+    looking up desktop files.
+    '''
+
+    def _on_received_app_infos(error, app_infos):
+        '''Called when we have our app_infos.
+
+        We can now marshal each app_info into an 'application' source
+        and continue using it that way.
+        '''
+        if error is not None:
+            callback(error, None)
+            return
+
+        callback(None, [
+            {
+                'type': 'application',
+                'detail': {
+                    'applicationId': a.app_id,
+                    'displayName': a.display_name,
+                    'shortDescription': a.short_description,
+                    'icon': format_app_icon_uri(a.icon, query['deviceUUID'])
+                }
+            }
+            for a in app_infos
+        ])
+
+    app_infos_for_feed_models(models, _on_received_app_infos)
+
+
+_SOURCE_IDENTIFIER_KEYS_BY_TYPE = {
+    'application': 'applicationId'
+}
+
+
+def get_source_identifier(source):
```

Comment:
This whole machinery feels quite awkward, since there is only one type and one identifier key...

Reply:
It is, though I'm designing for future sensibility here (we will eventually have card sources that are not necessarily applications and I don't want to start encoding that association now when I know it will have to change later).